spec/services/work_packages/set_attributes_service/derive_progress_values_work_based_spec.rb

```diff
@@ -79,45 +79,52 @@
       work_package.clear_changes_information
     end
 
-    context "when work is unset" do
+    context "when work is cleared" do
       let(:set_attributes) { { estimated_hours: nil } }
       let(:expected_derived_attributes) { { remaining_hours: nil } }
       let(:expected_kept_attributes) { %w[done_ratio] }
 
-      include_examples "update progress values", description: "keeps % complete, and unsets remaining work"
+      include_examples "update progress values", description: "keeps % complete, and clears remaining work"
     end
 
-    context "when remaining work is unset" do
+    context "when remaining work is cleared" do
       let(:set_attributes) { { remaining_hours: nil } }
-      let(:expected_derived_attributes) { { done_ratio: nil } }
-      let(:expected_kept_attributes) { %w[estimated_hours] }
+      let(:expected_derived_attributes) { { estimated_hours: nil } }
+      let(:expected_kept_attributes) { %w[done_ratio] }
 
-      include_examples "update progress values", description: "keeps work, and unsets % complete"
+      include_examples "update progress values", description: "keeps % complete, and clears work"
```

Comment:
Actually this is consistent. Well... kind-of.

In the initial implementation, some errors were displayed when all 3 fields were modified and the values were not consistent. Parimal objected that if we enter two values, then we should compute the third one, and discard the value entered by the user. Some values are more important than other: we try to keep the value of Work if it's entered, or % Complete.

Internally it tries to keep track of which values have been modified by the user.

So about you scenario, you start with all values set
Then you clear Remaining work. As a result, either % Complete or Work should be cleared. We debated this and decided that it made more sense to clear Work because they act together. That's why Work is cleared too.
Then you enter a value for Work. As a result, Remaining work is calculated and replaces the value entered.
Then you clear Remaining work again. And again, either % Complete or Work should be cleared. And this time, Work has been entered by the user which gives more importance to this value. So it's the value of % Complete which is cleared.

We can also see it like this: you clear Remaining work, and Work gets cleared as well. You think "No that's not what I want" and you put the value back in Work. Then you delete Remaining work and % Complete gets cleared. And that's what you wanted initially: keeping only Work for some reason.

Does it make more sense?

I agree that these scenarios are complicated and the logic is brittle. Other choices could have been made. That's why we documented them in the work package with all test cases.

Reply:
And about this scenario in the test case, it is correct: this is the first part of your test: 
- All values are set, when remaining work is cleared => keeps % complete, and clears work

Then in the second part, it's different as the Work is being set as well Remaining work is cleared. Both information are sent, and it always start from the values from the database (the server is stateless. The state is held in the frontend):
- All values are set, when work is changed to the same value and remaining work is cleared => % complete is cleared.

This is the spec example from line 310.

Reply:
It should also be easier to understand with the hints (from the other PR):

[demo with hints.webm](https://github.com/user-attachments/assets/2b4f5e69-1dc2-4da0-86c5-d93528cef73e)

Reply:
Thank you for the thorough explanation, it makes sense.